File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/lesson/LessonMapper.java
package kh.farrukh.sortmapperexamplejava.lesson;

import kh.farrukh.sortmapperexamplejava.course.CourseMapper;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonCreateRequestDTO;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonDetailsResponseDTO;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonResponseDTO;
import kh.farrukh.sortmapperexamplejava.lesson.model.LessonUpdateRequestDTO;
import kh.farrukh.sortmapperexamplejava.teacher.TeacherMapper;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants.ComponentModel;
import org.mapstruct.MappingTarget;

@Mapper(
    componentModel = ComponentModel.SPRING,
    uses = {CourseMapper.class, TeacherMapper.class}
)
public interface LessonMapper {

  @Mapping(target = "courseId", source = "course.id")
  @Mapping(target = "courseTitle", source = "course.title")
  @Mapping(target = "teacherFirstName", source = "course.teacher.firstName")
  @Mapping(target = "teacherLastName", source = "course.teacher.lastName")
  @Mapping(target = "helperTeacherId", source = "helperTeacher.id")
  @Mapping(target = "helperTeacherFirstName", source = "helperTeacher.firstName")
  @Mapping(target = "helperTeacherLastName", source = "helperTeacher.lastName")
  LessonResponseDTO toResponseDTO(LessonEntity entity);

  LessonDetailsResponseDTO toDetailsResponseDTO(LessonEntity entity);

  LessonEntity toEntity(LessonCreateRequestDTO requestDTO);

  void update(@MappingTarget LessonEntity entity, LessonUpdateRequestDTO requestDTO);

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/lesson/LessonMapper.kt
package kh.farrukh.sortmapperexample.lesson

import kh.farrukh.sortmapperexample.course.CourseEntity
import kh.farrukh.sortmapperexample.course.toResponseDTO
import kh.farrukh.sortmapperexample.lesson.model.LessonCreateRequestDTO
import kh.farrukh.sortmapperexample.lesson.model.LessonDetailsResponseDTO
import kh.farrukh.sortmapperexample.lesson.model.LessonResponseDTO
import kh.farrukh.sortmapperexample.lesson.model.LessonUpdateRequestDTO
import kh.farrukh.sortmapperexample.teacher.TeacherEntity
import kh.farrukh.sortmapperexample.teacher.toResponseDTO

fun LessonCreateRequestDTO.toEntity(
    course: CourseEntity,
    helperTeacher: TeacherEntity?
) = LessonEntity(
    title = title,
    description = description,
    course = course,
    helperTeacher = helperTeacher
)

fun LessonEntity.toResponseDTO() = LessonResponseDTO(
    id = id!!,
    title = title,
    description = description,
    courseId = course.id!!,
    courseTitle = course.title,
    teacherFirstName = course.teacher.firstName,
    teacherLastName = course.teacher.lastName,
    helperTeacherId = helperTeacher?.id,
    helperTeacherFirstName = helperTeacher?.firstName,
    helperTeacherLastName = helperTeacher?.lastName
)

fun LessonEntity.toDetailsResponseDTO() = LessonDetailsResponseDTO(
    id = id!!,
    title = title,
    description = description,
    courseId = course.id!!,
    course = course.toResponseDTO(),
    helperTeacherId = helperTeacher?.id,
    helperTeacher = helperTeacher?.toResponseDTO()
)

fun LessonEntity.update(
    requestDTO: LessonUpdateRequestDTO,
    course: CourseEntity,
    helperTeacher: TeacherEntity?
) {
    title = requestDTO.title
    description = requestDTO.description
    this.course = course
    this.helperTeacher = helperTeacher
}